### DISCUSSION TITLE
Using FATROP for Direct Collocation

### DISCUSSION THREAD

**Discussion by EserSteffen** [casadi/casadi#3844] [General]
Hello,

I've just stumbled upon the inclusion of FATROP in Casadi 3.6.6 and have just tried it out. 

I'm using the low-level interface of Casadi. After some code adjustments, I got the auto structure detection to work for multiple shooting, supplying the equality option with True on all gap-closing constraints, and switching the sign on those.

Some of my problems solve better when using direct collocation, so I was interested in implementing that. But all the resources I've found on fatrop do multiple shooting.
So my first question is, does it make sense / is possible to combine fatrop with collocation, or is it really only geared toward multiple shooting?

Attached is the debug view I am getting. 
In this problem, I have 1 state, 1 control and two other optimization variables that appear in constraints (slack vars etc.)
I am using a collocation of 2nd order and have set equality=True only for the gap-closing constraints, not the equality constraints within the collocation (not sure if this is correct).
In row 0 and 3 are the collocation polynomial constraints, in rows 1/2 4/5 are other path constraints, and row 6 is the gap-closing constraint.

![image](https://github.com/user-attachments/assets/17bc0824-fe28-4e51-b4aa-020e194c0d43)

Would greatly appreciate any help on this. And thanks to Joris and Joel for continuously putting work into casadi.

Best regards
Steffen

**Comment by EserSteffen** [comment]
Okay maybe the post was a bit unnecessary, starting on a fresh morning helped and I got it to work. 

Just following error messages was enough. I collected all the constraints defined in the inner collocation loop in a separate list, and added them to the constraint vector only AFTER the first gap-closing constraint. Which FATROP also states it expects the first constraint to be gap-closing. 

For anybody interested, this is what the new debug view looks like, MPC with this worked fine.
![image](https://github.com/user-attachments/assets/6622b380-fd71-4789-831f-2f73c0d2dbe2)

Also, of course I am lying on the equality vector, as the collocation polynomial constraints are equality constraints but I label them as false.

For me it works now, but if anybody with more knowledge on FATROP could confirm this makes sense would still be appreciated :) 

In my example, the multiple shooting version is dominated by Function Evaluation and I was not able to get a benefit from FATROP, but the collocation version benefitted hugely, giving a speedup of roughly 20 times.

**Reply by madcowswe** [reply]
@EserSteffen Do you mind sharing a snippet of code that has the speedup? Huge thanks!

I already have a working 2nd order Hermite Simpson direct collocation implementation inspired from "Collocation Methods for Second Order Systems"
And I'm now looking for ways to make it run fast.

**Reply by EserSteffen** [reply]
Hi, sorry for the late reply, I just saw this in my emails.

My code is part of the AgentLib library, you can see a collocation implementation here:
- https://github.com/RWTH-EBC/AgentLib-MPC/blob/main/agentlib_mpc/optimization_backends/casadi_/full.py#L43
- https://github.com/RWTH-EBC/AgentLib-MPC/blob/main/agentlib_mpc/optimization_backends/casadi_/basic.py#L249

A main file which leads into this would for example be here:
- https://github.com/RWTH-EBC/AgentLib-MPC/blob/main/examples/one_room_mpc/physical/simple_mpc.py#L140

It's not a snippet though, maybe it's a bit tough to follow since it's library code with some boilerplate / overhead.
The collocation implementation itself is adapted from casadi's own collocation example.

As a caveat though, I have mainly moved away from fatrop currently, as I found result quality to be a bit inferior to IPOPT in some cases.

**Comment by ImW1ll** [comment]
Hi!
I'm trying to get it working also for collocation and it's kicking my butt. What type of integrator do you have: explicit or implicit? If you have implicit, may I ask how you made it work? Thanks!

**Reply by madcowswe** [reply]
I think you are on to something: typical trapezoidal and hermite simpson collocation formulations are doing the equivalent to implicit integration. Which seems unsupported, it seems like FATROP requires explicit integration only? I get this impression from the FATROP paper which writes the problem like x_k+1 = f_k(u_k, x_k).
Also the debug view seems to require that the link between x_k and x_k+1 is the identity matrix only (labeled `I` in the debug view).

I was able to hack my collocation formulation into "explicit" integration form by introducing extra "local" variables for the right hand nodes on each stage, and then writing the collocation formulation (implicit integration) with the local variables, and then extra gap closing to say that the right hand nodes = the next stage left nodes.

However this ran at about the same speed as ipopt did when using the original formulation, so doesn't seem like a good workaround.

```py
# Auxilary gap closing
opti.subject_to(q_var[:, k+1] == q_next_var[:, k])
opti.subject_to(v_var[:, k+1] == v_next_var[:, k])
opti.subject_to(u_var[k+1] == u_next_var[k])

# dynamics in terms of q, v, u, q_next, v_next, u_next
# (dynamics eqns)
```
<img width="906" height="784" alt="image" src="https://github.com/user-attachments/assets/20b9c98e-5a92-463c-87af-b1adc5728f8f" />

**Reply by madcowswe** [reply]
Ok actually maybe the weird workaround could help if you also do jit/code gen, since it ended up dumping a bunch of time over to function evaluation instead. I guess also due to the indirection in the formulation, it takes ~50% more iterations:
```
               Solver  FE time (ms)  Rest time (ms)  Total time (ms)   Iterations
0    IPOPT (straight)         6.694          15.896            22.59   11
1 FATROP (workaround)        22.759           2.621            25.38   15
```

Still feels a bit uncomfortable to build a solution with this workaround. I wonder if there is scope to support implicit integration in FATROP at some point (or even better if I misunderstood and it's actually supported)

**Reply by ImW1ll** [reply]
So sorry for not answering earlier @madcowswe. I'm back on a similar project almost 2 years after and this helped me haha
Thanks so much!

**Comment by EserSteffen** [comment]
Hi, if this is still relevant, I quickly created a comparison for collocation, using the van der pol example from casadi's example pack as a base.
Speedup in this example is around x3.

https://gist.github.com/EserSteffen/1ff95d52c6ccf02b348f1fd5821609fb

I have also attached a screenshot of the diff view to see exactly what is changed between the two versions.

<img width="1422" height="1190" alt="image" src="https://github.com/user-attachments/assets/5cadeb37-dfd7-413c-ad0d-c5d7f447e727" />
